### ISSUE TITLE
Out of memory error specifically when using a mask on a 16bit stack in FindFoci

### ISSUE THREAD

**Issue by lollopus** [aherbert/GDSC#4]
I've followed your advice and solved the out of memory problems reported in https://github.com/aherbert/gdsc/issues/2 by scaling my stack from 32 down to 16 bits. However, as soon as I attempted to restrict the analysed volume by using a mask I ran into an out of memory error again.

This surprised me, since according to your detailed explanation my 1.9 GB 16bit input stack should use only 9GB out of the 21GB made available to Fiji. Moreover, if I understood correctly you use masks only to decide which peaks to report to the user. So I can't understand why adding a mask would require an additional 12GB of Ram for processing. I thus suspect this may be a bug!

To summarise:

Input stack 1414x5796x125px, 16 bit, 1.9GB
>FindFoci runs without issues

Input stack: 1414x5796x125px, 16 bit, 1.9GB + Mask: 1414x5796x125px, 8 bit, 0.9GB (only 0 and 255 as values, with 255 in a small square volume)
>FindFoci runs out of memory immediately (see below)

```
FindFoci : LoG, scaled 0-300 to 16bit of female, fused.tif
Mask image = mask2.tif
Image min = 0
Image stats (inside mask/ROI) : Min = 0, Max = 45075, Mean = 2622.63, StdDev = 603.27
Background stats (mode=Both) : Min = 0, Max = 65535, Mean = 2160.38, StdDev = 1504.70
Initialised = 4182.86 ms : 4182.86 ms
Background level = 4000
<Out of memory>

```

**Comment by lollopus** [comment]
This is the output of the error handler, hope it helps.

```
(Fiji Is Just) ImageJ 2.0.0-rc-68/1.52h29; Java 1.8.0_66 [64-bit]; Mac OS X 10.13.6; 11765MB of 20444MB (57%)
 
java.lang.OutOfMemoryError: Java heap space
	at uk.ac.sussex.gdsc.foci.FindFociOptimisedIntProcessor.getSortedMaxPoints(FindFociOptimisedIntProcessor.java:192)
	at uk.ac.sussex.gdsc.foci.FindFociBaseProcessor.findMaxima(FindFociBaseProcessor.java:246)
	at uk.ac.sussex.gdsc.foci.FindFoci.findMaxima(FindFoci.java:1836)
	at uk.ac.sussex.gdsc.foci.FindFoci.exec(FindFoci.java:1001)
	at uk.ac.sussex.gdsc.foci.FindFoci.run(FindFoci.java:832)
	at ij.IJ.runUserPlugIn(IJ.java:228)
	at ij.IJ.runPlugIn(IJ.java:192)
	at ij.Executer.runCommand(Executer.java:137)
	at ij.Executer.run(Executer.java:66)
	at java.lang.Thread.run(Thread.java:745)

```

**Comment by aherbert** [comment]
I have updated the plugin again. See if that helps.

**Comment by lollopus** [comment]
Problem solved it seems. I could run the same analysis with _Maximum peaks_ = 10000

Thanks for fixing this!

**Comment by aherbert** [comment]
Thanks for letting me know. I hope to try and implement better memory handling for crops soon.
